Remove NEXUS ROOT Dependence

kvjmistry opened the pull request:
This PR removes the ROOT dependence in NEXUS, primarily in the Muon Generators. Now, events are generated using a csv file rather than a root file. 

An additional simulated muon file is included which is generated from the MCEq and Proposal simulation. This can be chosen from the config macro file when running the generator. 

Tested compilation on Harvard Cluster (Cent Os) and Macbook M1. 

Please see https://next.ific.uv.es/cgi-bin/DocDB/private/ShowDocument?docid=1325 for slides explaining this PR.

kvjmistry committed:
Remove ROOT dependency of action and physics classes

kvjmistry committed:
Remove ROOT dependency from muon generators

kvjmistry committed:
Relocate and add muon distribution files

kvjmistry committed:
Update muon generation config files

kvjmistry committed:
Add catch for unphysical sampled zenith values

kvjmistry committed:
Update muon simulation file

kvjmistry committed:
Remove ROOT requirement in cmake and scons

kvjmistry committed:
Revert default muon angle config

kvjmistry committed:
Add README files for documentation

kvjmistry committed:
Fix CSV file exception in MuonAngleGenerator

gondiaz commented on source/actions/MuonsEventAction.cc:
```diff
@@ -21,10 +21,8 @@
 #include <G4SDManager.hh>
 #include <G4HCtable.hh>
```
`G4HCtable.hh` and `G4HCofThisEvent.hh` are not needed

gondiaz commented on source/actions/ValidationTrackingAction.cc:
```diff
@@ -23,9 +23,8 @@
 #include <G4ParticleDefinition.hh>
```
`G4ParticleDefinition.hh` not needed

kvjmistry replied:
Thanks, I have removed the headers.

kvjmistry committed:
Remove unused header includes

gondiaz commented on source/actions/MuonsEventAction.cc:
```diff
@@ -18,13 +18,9 @@
 #include <G4Trajectory.hh>
 #include <G4GenericMessenger.hh>
 #include <G4HCofThisEvent.hh>
```
Remove `G4HCofThisEvent.hh` also

gondiaz commented on source/generators/MuonAngleGenerator.h:
```diff
@@ -71,12 +78,22 @@ namespace nexus {
+    std::mt19937 RN_engine_az; ///< Random Number Generator Engine Azimuth
+    std::mt19937 RN_engine_zen; ///< Random Number Generator Engine Zenith
+
+    std::discrete_distribution<G4int> discr_dist; ///< Discrete distribution to sample flux
```
By convention, data members in nexus are followed by a trailing underscore, see the [Coding conventions and style guide](https://github.com/next-exp/nexus/wiki/Coding-conventions-and-style-guide)

kvjmistry replied:
Thanks for bringing this to my attention, I was not aware of this practice and will update the variable names to match this style.

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -110,6 +164,18 @@ void MuonAngleGenerator::SetupAngles()
 {
 
+  // Initalise RNG seeds and load file on the first event
+  if (event->GetEventID() == 0){ 
```
I think we must avoid this clause. Notice that the first event id is not 0 in general, since this is set via `/nexus/persistency/start_id`. What would the seed be in such a case? Could we just take the seed from `/nexus/random_seed`?

kvjmistry replied:
Right, I couldn't figure out a way to initialize it only once, this protection was added to stop initializing it on every new event. We could possibly use the same protection as the rotation is initialised ie. if ( rPhi_ == NULL) instead if you prefer.  

The CLHEP::HepRandom::getTheSeed() should be already grabbing the /nexus/random_seed configured.

kvjmistry replied:
or we can add a data member bool to flag if initialised, that is a little clearer actually.

jmalbos replied:
Do we really need those `RN_engine`? Can't you just use the functions (e.g. `G4UniformRandom()`) provided by Geant4?

https://geant4.kek.jp/lxr/source/global/HEPRandom/include/Randomize.hh

jmalbos replied:
> I think we must avoid this clause. Notice that the first event id is not 0 in general, since this is set via `/nexus/persistency/start_id`. 

Just for the sake of correctness: The first event ID is always 0. We only modify it at the time we write the events in the output file.

kvjmistry replied:
The random engine is used as an input to the std::discrete_distribution which is what I use to sample the events from. I don't think I can use the G4UniformRandom() for that. 

I can probably replace the other RN generators used for the Gaussian smearing, but I just chose to use the std functions for consistency. Let me know if you would like to use the Geant4 function in this case and I can swap it.

kvjmistry committed:
Fix variable names to match code conventions

kvjmistry committed:
Update init clause in Muon Generator

kvjmistry committed:
Update simulated muon angular file

kvjmistry committed:
Fix Geant4 Ntuple in MuonEventAction

gondiaz commented on source/actions/MuonsEventAction.cc:
```diff
@@ -17,14 +17,9 @@
-#include <G4SDManager.hh>
-#include <G4HCtable.hh>
 #include <globals.hh>
+#include "g4root_defs.hh"
```
Is this header necessary?

gondiaz commented on source/actions/ValidationTrackingAction.cc:
```diff
@@ -20,12 +20,10 @@
-#include <G4ParticleDefinition.hh>
 #include <G4OpticalPhoton.hh>
 #include <G4Gamma.hh>
+#include "g4root_defs.hh"
```
Same as above

kvjmistry replied:
These headers pull in the geant4 histogram and ntuples that we are replacing the ROOT histograms with. To keep with convention, I should actually use <> rather than "" in these include lines.

gondiaz replied:
Histograms and NTuples are created through G4AnalysisManager methods [Analysis Manager Classes](https://geant4-userdoc.web.cern.ch/UsersGuides/ForApplicationDeveloper/html/Analysis/managers.html#files-handling). Therefore whatever in this include is not been used.
The manager chooses the default output file from the filename extension (.csv, .hdf5, .root or .xml). I have tried .csv, .root and .xml and seems to work. The .hdf5 should work if Geant4 is compiled against hdf5.

kvjmistry replied:
Yep your right, I removed the superfluous header.

gondiaz reviewed (commented):
Review includes in MuonsEventAction.cc, some of them are not needed

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
+    // Calculate the vector components of the muon
+    dir.setX(sin(zen_samp) * sin(az_samp));
+    dir.setY(-cos(zen_samp));
+    dir.setZ(-sin(zen_samp) * cos(az_samp));
```
This assumes that azimuth is the angle between muon direction and -Z axis. Is it the intented behaviour? I would naively consider the azimuth to be between muon direction and +X. 
Also, which is the value of `azimuth_rotation` for NEXT100? I see that `NEXT100_muons_lsc` and `NEXT100_muons_hallA` use the value of 150 degrees. The rotation here is performed anticlockwise throughout Y axis (meaning -150). Is it the correct orientation?

andLaing replied:
All these slightly non-intuitive definitions are due to the difference in axis orientation between NEXT, set so that positive z points towards the cathode, and the 'natural' orientation where positive z points upwards. Additionally, the measurements of the flux in the lab are oriented so that the azimuthal angle 0 is at north so we need to have the azimuth angle around Y (which points up in next) and rotate the axes in the input info to the orientation in NEXT. Since NEXT100 will be in the same shielding as NEW (unless things have changed since I was in NEXT) we always assumed the 150 degrees rotation of the axes would be the same but it's a configuration variable in case of changes there.

There's PR, that's probably very out of date, which updates a lot of important things for muon generation (#69 , think) to make it closer to reality and to aid normalisation, maybe worthwhile having a look.

Hope that helps.

gondiaz replied:
Thanks for the clarification!

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -110,6 +164,21 @@ void MuonAngleGenerator::SetupAngles()
 void MuonAngleGenerator::GeneratePrimaryVertex(G4Event* event)
```
`MuonAngleGenerator` breaks when used with `MuonsEventAction`. Add the angles to the userInfo (check how it is done in`MuonGenerator`)

kvjmistry replied:
Added angles to the user info.

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
+          }
+        }
+    
+    }
```
Assuming all bins are in the input file, I would suggest to simplify the code to (same for azimuth smearing)

```suggestion
    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
          if (zeniths_[RN_indx] >= zenith_bins_[i] 
              && zeniths_[RN_indx] < zenith_bins_[i+1]){
              zen_BW_smear = zen_BW_[i];
              break;
          }
```

Also, I would set the default `zen_BW_smear` and `az_BW_smear` to a meaninless negative value, say -1.

kvjmistry replied:
Tidied up code by including a new function. Also set default values from 1.0e6 to std::numeric_limits<G4double>::lowest().

Kept two if statements, still need to keep the very last bin with the additional <= sign so it is included.

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
+    // Check if the smear values are set properly
+    if (az_BW_smear == 1.0e6 || zen_BW_smear == 1.0e6 )
+      G4Exception("[MuonAngleGenerator]", "GetDirection()",
+                FatalException, " bin smear value was not set correctly ");
```
Following previous comment, if smear values are -1, it would mean that the sampled value (`azimuths_[RN_indx], zeniths_[RN_indx]`) is not inside any of the bins in the input file. Therefore I think a message such that "Sampled angle not inside binning, review input file" will be helpfull.

gondiaz replied:
We might also avoid doing this check for every sampling, and add a check above such that
`azimuth_bins_[0] <= azimuths_.min() && azimuths_.max() <=  azimuth_bins_[-1]`. This also avoids to define meaninless `az_BW_smear`. Same for zenith

kvjmistry replied:
Updated error message to be more descriptive. 

I do not quite understand what you mean by the second check here, happy to follow up with you on slack about this.

gondiaz replied:
Defining a meaningless `smearVal` is only used to check is the sampled `sampVal` is inside any of the bins provided at the configuration file, explicitly:

```
    // Check if the smear values are set properly
    if (az_BW_smear == std::numeric_limits<G4double>::lowest() ||
        zen_BW_smear == std::numeric_limits<G4double>::lowest() ){
                G4Exception("[MuonAngleGenerator]", "GetDirection()",
                FatalException, " Sampled angle not inside binning, review input file ");
    }
```

My point is that instead doing this check at each call to `GetDirection`, doing it globally in `LoadMuonDistribution` with something like (pseudocode):
`azimuth_bins_[0] <= min(azimuths_) && max(azimuths_) <= azimuth_bins_[-1]`

Also, we must be sure that `azimuth_bins_` are sorted, then I would suggest to sort them once they are pulled from the input file. 
Sure, contact me if I am not explaining myself!

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
+    std::normal_distribution<G4double> Gauss_zen(0, zen_BW_smear);
+    
+    // Get the Gaussian smear values 
+    G4double az_smear  = Gauss_az(RN_engine_az_);
```
As commented above, I would use G4 builtins, in this case `G4RandGauss::shoot`

kvjmistry replied:
Updated generator to use the CLHEP RandGauss tools rather than the std library.

gondiaz replied:
I would avoid defining two extra random engines but use `G4RandGauss::shoot` instead, what do you think @jmalbos @paolafer?

kvjmistry committed:
Add angles in MuonAngleGenerator to the user info

Remove superfluous header

Update default unset zenith and azimuth to more unphyical value

kvjmistry committed:
Add binSmear function to reduce repeated code

kvjmistry committed:
Replace Gaussian RNG with CLHEP tools

Fix typo

kvjmistry committed:
Use replace CLHEP random generator with G4RandGauss:shoot

kvjmistry committed:
Use G4RandGeneral function to sample muon flux

kvjmistry committed:
Move check for angles to muon initalisation

Fix typo

kvjmistry committed:
Add energy input to MuonAngleGenerator

Update muon angle and energy files

Fix pi conversion factor in muon generator files

Remove unused namespace

Remove unused headers

Put back removed header include

gondiaz reviewed (commented):
Are the smear values in the `MuonAnaAllRuns.csv` distribution file correct?
The azimuth bin-widths are ~0.15, computed from the bins, but the azimuth_smear is always 0.05. For zenith the values are ~0.039 and 0.0125.

gondiaz commented on data/README.md:
```diff
@@ -0,0 +1,14 @@
+Bi214.genbb: File for generating Bi214 events  
+Tl208.genbb: File for generating Tl208 events  
+Xe136_bb0nu.genbb: File for generating 0nu double beta events  
+Xe136_bb2nu.genbb: File for generating 2nu double beta events  
```
I don't think this is correct, they seem generated instead of generation files...

gondiaz commented on source/generators/MuonGenerator.cc:
```diff
@@ -21,12 +21,8 @@
-#include <TF1.h>
-#include <TMath.h>
-
 #include "CLHEP/Units/SystemOfUnits.h"
```
not needed, same for `using namespace CLHEP;` below

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -39,7 +35,7 @@ REGISTER_CLASS(MuonAngleGenerator, G4VPrimaryGenerator)
   G4VPrimaryGenerator(), msg_(0), particle_definition_(0),
   angular_generation_(true), rPhi_(NULL), energy_min_(0.),
-  energy_max_(0.), distribution_(0), geom_(0), geom_solid_(0)
+  energy_max_(0.), geom_(0), geom_solid_(0), bInitialize_(false)
```
Add default value of `dist_name_` to "za". If it is not set the code breaks without any hint about the reason.

gondiaz commented on source/generators/MuonAngleGenerator.cc:
```diff
@@ -25,10 +25,6 @@
-#include <TFile.h>
-#include <TH2F.h>
-
 #include "CLHEP/Units/SystemOfUnits.h"
```
not needed, same for `using namespace CLHEP;` below

gondiaz reviewed (commented):
For consistency, remove execution file permissions in SimulatedMuonsProposalMCEqEnergy.csv.

gondiaz reviewed (approved):
This PR replaces the ROOT dependency on nexus, which was present in many parts of the code. The most relevant feature is the substitution of ROOT's `TFH` histograms by the Geant4's `G4RandGeneral` sampling function. This requires the change of the previously used root format to csv format. The PR also adds the muon flux files from simulations.
This PR overcomes #55, providing a long ago desired fix. Great work!
